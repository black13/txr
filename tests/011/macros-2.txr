@(do
   (defmacro while ((condition : result) . body)
     (let ((cblk (gensym "cnt-blk-"))
           (bblk (gensym "brk-blk-")))
       ^(macrolet ((break (value) ^(return-from ,',bblk ,value)))
          (symacrolet ((break (return-from ,bblk))
                       (continue (return-from ,cblk)))
            (block ,bblk
              (for () (,condition ,result) ()
                (block ,cblk ,*body)))))))

   (let ((i 0))
     (while ((< i 100))
       (if (< (inc i) 20)
         continue)
       (if (> i 30)
         break)
       (prinl i)))

   (prinl
     (sys:expand
       '(while ((< i 100))
          (if (< (inc i) 20)
            continue)
          (if (> i 30)
            break)
          (prinl i)))))
