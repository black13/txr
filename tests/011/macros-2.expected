20
21
22
23
24
25
26
27
28
29
30
(block #:brk-blk-0006 (for nil ((< i 100) nil) nil (block #:cnt-blk-0005 (if (< (inc i) 20) (return-from #:cnt-blk-0005)) (if (> i 30) (return-from #:brk-blk-0006)) (prinl i))))
(while break)
(while break)
(while break)
(while break)
(while break)
